IMPORT LIBRARY

In [69]:
import pandas as pd
import matplotlib.pyplot as plt
#%matplotlib inline
import seaborn as sns

In [70]:
!pip install seaborn

In [71]:
!pip install matplotlib

IMPORT FILE CSV

In [72]:
df = pd.read_csv('ds_salaries.csv')

In [73]:
df.head()

work_year experience_level employment_type                 job_title  \
0       2023               SE              FT  Principal Data Scientist   
1       2023               MI              CT               ML Engineer   
2       2023               MI              CT               ML Engineer   
3       2023               SE              FT            Data Scientist   
4       2023               SE              FT            Data Scientist   

   salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
0   80000             EUR          85847                 ES           100   
1   30000             USD          30000                 US           100   
2   25500             USD          25500                 US           100   
3  175000             USD         175000                 CA           100   
4  120000             USD         120000                 CA           100   

  company_location company_size  
0               ES            L  
1               US            S  
2               US            S  
3               CA            M  
4               CA            M

# 1 Exploratory Data Analysis




## 1.1 Check Null Values

In [74]:
df.isna().sum()

work_year             0
experience_level      0
employment_type       0
job_title             0
salary                0
salary_currency       0
salary_in_usd         0
employee_residence    0
remote_ratio          0
company_location      0
company_size          0
dtype: int64

## 1.2 Check Duplicate Data

In [75]:
df.shape

(3755, 11)

In [76]:
df[df.duplicated()]

#data tanpa identifier (primary key kemungkinan besar ada duplikatnya)

work_year experience_level employment_type           job_title  salary  \
115        2023               SE              FT      Data Scientist  150000   
123        2023               SE              FT  Analytics Engineer  289800   
153        2023               MI              FT       Data Engineer  100000   
154        2023               MI              FT       Data Engineer   70000   
160        2023               SE              FT       Data Engineer  115000   
...         ...              ...             ...                 ...     ...   
3439       2022               MI              FT      Data Scientist   78000   
3440       2022               SE              FT       Data Engineer  135000   
3441       2022               SE              FT       Data Engineer  115000   
3586       2021               MI              FT       Data Engineer  200000   
3709       2021               MI              FT      Data Scientist   76760   

     salary_currency  salary_in_usd employee_residence  remote_ratio  \
115              USD         150000                 US             0   
123              USD         289800                 US             0   
153              USD         100000                 US           100   
154              USD          70000                 US           100   
160              USD         115000                 US             0   
...              ...            ...                ...           ...   
3439             USD          78000                 US           100   
3440             USD         135000                 US           100   
3441             USD         115000                 US           100   
3586             USD         200000                 US           100   
3709             EUR          90734                 DE            50   

     company_location company_size  
115                US            M  
123                US            M  
153                US            M  
154                US            M  
160                US            M  
...               ...          ...  
3439               US            M  
3440               US            M  
3441               US            M  
3586               US            L  
3709               DE            L  

[1171 rows x 11 columns]

In [77]:
df[df.duplicated()].shape

(1171, 11)

### Copy Data

In [78]:
# copy data digunakan agar data aslinya utuh
# karena nantinya data data duplikat akan di drop

df_2 = df.copy()

In [79]:
df_2.shape

(3755, 11)

In [80]:
pd.options.mode.chained_assignment = None

### Delete Duplicate Data

In [81]:
df_2 = df_2.drop_duplicates()

In [82]:
df_2.shape

(2584, 11)

## 1.3 Check Unique Values in Each Column

### Check Data Type

In [83]:
df_2.dtypes

work_year              int64
experience_level      object
employment_type       object
job_title             object
salary                 int64
salary_currency       object
salary_in_usd          int64
employee_residence    object
remote_ratio           int64
company_location      object
company_size          object
dtype: object

### Check Data stats

In [84]:
df_2.describe()

work_year        salary  salary_in_usd  remote_ratio
count  2584.000000  2.584000e+03    2584.000000   2584.000000
mean   2022.301084  2.103653e+05  133409.280186     50.483746
std       0.749179  8.080375e+05   67136.837329     48.163707
min    2020.000000  6.000000e+03    5132.000000      0.000000
25%    2022.000000  9.000000e+04   84975.000000      0.000000
50%    2022.000000  1.346300e+05  130000.000000     50.000000
75%    2023.000000  1.825625e+05  175000.000000    100.000000
max    2023.000000  3.040000e+07  450000.000000    100.000000

### Check Columns

In [122]:
df_2.columns

Index(['work_year', 'experience_level', 'employment_type', 'job_title',
       'salary', 'salary_currency', 'salary_in_usd', 'employee_residence',
       'remote_ratio', 'company_location', 'company_size', 'job_category',
       'adjusted_salary'],
      dtype='object')

### Unique Valeus Check

In [121]:
for column in df_2.columns:
    if df_2[column].dtypes == object:
        num_distinct = df_2[column].nunique()
        distinct_val = df_2[column].unique()
        print(f"{column} : {num_distinct} --> {distinct_val}\n")

experience_level : 4 --> ['Senior' 'Mid / Intermediete Level' 'Entry Level' 'Executive']

employment_type : 4 --> ['Full Time' 'Contract' 'Freelance' 'Part Time']

job_title : 93 --> ['Principal Data Scientist' 'ML Engineer' 'Data Scientist'
 'Applied Scientist' 'Data Analyst' 'Data Modeler' 'Research Engineer'
 'Analytics Engineer' 'Business Intelligence Engineer'
 'Machine Learning Engineer' 'Data Strategist' 'Data Engineer'
 'Computer Vision Engineer' 'Data Quality Analyst'
 'Compliance Data Analyst' 'Data Architect'
 'Applied Machine Learning Engineer' 'AI Developer' 'Research Scientist'
 'Data Analytics Manager' 'Business Data Analyst' 'Applied Data Scientist'
 'Staff Data Analyst' 'ETL Engineer' 'Data DevOps Engineer' 'Head of Data'
 'Data Science Manager' 'Data Manager' 'Machine Learning Researcher'
 'Big Data Engineer' 'Data Specialist' 'Lead Data Analyst'
 'BI Data Engineer' 'Director of Data Science'
 'Machine Learning Scientist' 'MLOps Engineer' 'AI Scientist'
 'Autonomous V

## 1.4 Rename Value for Better Understanding

In [87]:
df_2['experience_level'] = df_2['experience_level'].map({
    'SE' : 'Senior',
    'EN' : 'Entry Level',
    'EX' : 'Executive',
    'MI' : 'Mid / Intermediete Level'
})
df_2['employment_type'] = df_2['employment_type'].map({
    'FL' : 'Freelance',
    'CT' : 'Contract',
    'FT' : 'Full Time',
    'PT' : 'Part Time'
})
df_2['remote_ratio'] = df_2['remote_ratio'].map({
    0 : 'On-site',
    50 : 'Hybrid',
    100 : 'Remote'
})



In [120]:
df_2.head()

work_year          experience_level employment_type  \
0       2023                    Senior       Full Time   
1       2023  Mid / Intermediete Level        Contract   
2       2023  Mid / Intermediete Level        Contract   
3       2023                    Senior       Full Time   
4       2023                    Senior       Full Time   

                  job_title  salary salary_currency  salary_in_usd  \
0  Principal Data Scientist   80000             EUR          85847   
1               ML Engineer   30000             USD          30000   
2               ML Engineer   25500             USD          25500   
3            Data Scientist  175000             USD         175000   
4            Data Scientist  120000             USD         120000   

  employee_residence remote_ratio company_location company_size  \
0                 ES       Remote               ES            L   
1                 US       Remote               US            S   
2                 US       Remote               US            S   
3                 CA       Remote               CA            M   
4                 CA       Remote               CA            M   

       job_category  adjusted_salary  
0            Others          85847.0  
1  Machine Learning          30000.0  
2  Machine Learning          25500.0  
3    Data Scientist         175000.0  
4    Data Scientist         120000.0

In [119]:
#ditemukan, remote ratio yang Nan

df_2.isna().sum()

work_year             0
experience_level      0
employment_type       0
job_title             0
salary                0
salary_currency       0
salary_in_usd         0
employee_residence    0
remote_ratio          0
company_location      0
company_size          0
job_category          0
adjusted_salary       0
dtype: int64

In [118]:
df_2['remote_ratio'] = df_2['remote_ratio'].fillna(df_2['remote_ratio'].mode()[0])
df_2.isna().sum()

# Nan hilang diisi dengan modus

work_year             0
experience_level      0
employment_type       0
job_title             0
salary                0
salary_currency       0
salary_in_usd         0
employee_residence    0
remote_ratio          0
company_location      0
company_size          0
job_category          0
adjusted_salary       0
dtype: int64

## 1.5 Group the Job title

### Check Job Title

In [117]:
print(df_2['job_title'].nunique())
print(df_2['job_title'].unique())

93
['Principal Data Scientist' 'ML Engineer' 'Data Scientist'
 'Applied Scientist' 'Data Analyst' 'Data Modeler' 'Research Engineer'
 'Analytics Engineer' 'Business Intelligence Engineer'
 'Machine Learning Engineer' 'Data Strategist' 'Data Engineer'
 'Computer Vision Engineer' 'Data Quality Analyst'
 'Compliance Data Analyst' 'Data Architect'
 'Applied Machine Learning Engineer' 'AI Developer' 'Research Scientist'
 'Data Analytics Manager' 'Business Data Analyst' 'Applied Data Scientist'
 'Staff Data Analyst' 'ETL Engineer' 'Data DevOps Engineer' 'Head of Data'
 'Data Science Manager' 'Data Manager' 'Machine Learning Researcher'
 'Big Data Engineer' 'Data Specialist' 'Lead Data Analyst'
 'BI Data Engineer' 'Director of Data Science'
 'Machine Learning Scientist' 'MLOps Engineer' 'AI Scientist'
 'Autonomous Vehicle Technician' 'Applied Machine Learning Scientist'
 'Lead Data Scientist' 'Cloud Database Engineer' 'Financial Data Analyst'
 'Data Infrastructure Engineer' 'Software Data Eng

### Grouping Function

In [92]:
def assign_broader_category(job_title):
    data_engineering = ["Data Engineer", "Data Analyst", "Analytics Engineer", "BI Data Analyst", "Business Data Analyst", "BI Developer", "BI Analyst", "Business Intelligence Engineer", "BI Data Engineer", "Power BI Developer"]
    data_scientist = ["Data Scientist", "Applied Scientist", "Research Scientist", "3D Computer Vision Researcher", "Deep Learning Researcher", "AI/Computer Vision Engineer"]
    machine_learning = ["Machine Learning Engineer", "ML Engineer", "Lead Machine Learning Engineer", "Principal Machine Learning Engineer"]
    data_architecture = ["Data Architect", "Big Data Architect", "Cloud Data Architect", "Principal Data Architect"]
    management = ["Data Science Manager", "Director of Data Science", "Head of Data Science", "Data Scientist Lead", "Head of Machine Learning", "Manager Data Management", "Data Analytics Manager"]
    if job_title in data_engineering:
        return "Data Engineering"
    elif job_title in data_scientist:
        return "Data Scientist"
    elif job_title in machine_learning:
        return "Machine Learning"
    elif job_title in data_architecture:
        return "Data Architecture"
    elif job_title in management:
        return "Management"
    else:
        return "Others"
    
           

In [116]:
df_2['job_category'] = df_2['job_title'].apply(assign_broader_category)

In [115]:
df_2.head()

work_year          experience_level employment_type  \
0       2023                    Senior       Full Time   
1       2023  Mid / Intermediete Level        Contract   
2       2023  Mid / Intermediete Level        Contract   
3       2023                    Senior       Full Time   
4       2023                    Senior       Full Time   

                  job_title  salary salary_currency  salary_in_usd  \
0  Principal Data Scientist   80000             EUR          85847   
1               ML Engineer   30000             USD          30000   
2               ML Engineer   25500             USD          25500   
3            Data Scientist  175000             USD         175000   
4            Data Scientist  120000             USD         120000   

  employee_residence remote_ratio company_location company_size  \
0                 ES       Remote               ES            L   
1                 US       Remote               US            S   
2                 US       Remote               US            S   
3                 CA       Remote               CA            M   
4                 CA       Remote               CA            M   

       job_category  adjusted_salary  
0            Others          85847.0  
1  Machine Learning          30000.0  
2  Machine Learning          25500.0  
3    Data Scientist         175000.0  
4    Data Scientist         120000.0

## 1.6 Adjust Salary to Present Values

In [114]:
df_2['work_year'].unique()

array([2023, 2022, 2020, 2021])

### Inflation Rate

In [96]:
# Inflation rates
us_inflation_rates = {2019: 0.0181,
                      2020: 0.0123,
                      2021: 0.0470,
                      2022: 0.065}
global_inflation_rates = {2019: 0.0219,
                          2020: 0.0192,
                          2021: 0.0350,
                          2022: 0.088}

def adjust_salary(x):
    year = x ['work_year']
    ori_salary = x['salary_in_usd']
    currency = x['salary_currency']
    
    if year == 2023:
        return ori_salary
    else:
        adjust_salary = ori_salary
        
        for yr in range(year,2023):
            if currency == 'USD':
                inflation_rate = us_inflation_rates[yr]
            else:
                inflation_rate = global_inflation_rates[yr]
        adjust_salary = adjust_salary * (1 + inflation_rate)
    return adjust_salary
            

In [113]:
df_2['adjusted_salary'] = df_2.apply(adjust_salary, axis =1)

In [112]:
df_2[['work_year', 'salary_currency', 'salary_in_usd','adjusted_salary']].tail()

work_year salary_currency  salary_in_usd  adjusted_salary
3750       2020             USD         412000        438780.00
3751       2021             USD         151000        160815.00
3752       2020             USD         105000        111825.00
3753       2020             USD         100000        106500.00
3754       2021             INR          94665        102995.52

In [99]:
df_2.head()

work_year          experience_level employment_type  \
0       2023                    Senior       Full Time   
1       2023  Mid / Intermediete Level        Contract   
2       2023  Mid / Intermediete Level        Contract   
3       2023                    Senior       Full Time   
4       2023                    Senior       Full Time   

                  job_title  salary salary_currency  salary_in_usd  \
0  Principal Data Scientist   80000             EUR          85847   
1               ML Engineer   30000             USD          30000   
2               ML Engineer   25500             USD          25500   
3            Data Scientist  175000             USD         175000   
4            Data Scientist  120000             USD         120000   

  employee_residence remote_ratio company_location company_size  \
0                 ES       Remote               ES            L   
1                 US       Remote               US            S   
2                 US       Remote               US            S   
3                 CA       Remote               CA            M   
4                 CA       Remote               CA            M   

       job_category  adjusted_salary  
0            Others          85847.0  
1  Machine Learning          30000.0  
2  Machine Learning          25500.0  
3    Data Scientist         175000.0  
4    Data Scientist         120000.0

# SOAL No.1
Show the median adjusted data scientist salary, for each experience level

## JAWABAN No.1

In [100]:
# filter job_tite ; data scientist
df_science = df_2[df_2['job_title'] == 'Data Scientist']

In [101]:
df_science

work_year          experience_level employment_type       job_title  \
3          2023                    Senior       Full Time  Data Scientist   
4          2023                    Senior       Full Time  Data Scientist   
7          2023                    Senior       Full Time  Data Scientist   
8          2023                    Senior       Full Time  Data Scientist   
9          2023                    Senior       Full Time  Data Scientist   
...         ...                       ...             ...             ...   
3741       2020  Mid / Intermediete Level       Full Time  Data Scientist   
3742       2020  Mid / Intermediete Level       Full Time  Data Scientist   
3746       2021  Mid / Intermediete Level       Full Time  Data Scientist   
3750       2020                    Senior       Full Time  Data Scientist   
3752       2020               Entry Level       Full Time  Data Scientist   

      salary salary_currency  salary_in_usd employee_residence remote_ratio  \
3     175000             USD         175000                 CA       Remote   
4     120000             USD         120000                 CA       Remote   
7     219000             USD         219000                 CA      On-site   
8     141000             USD         141000                 CA      On-site   
9     147100             USD         147100                 US      On-site   
...      ...             ...            ...                ...          ...   
3741  118000             USD         118000                 US       Remote   
3742  138350             USD         138350                 US       Remote   
3746  160000             SGD         119059                 SG       Remote   
3750  412000             USD         412000                 US       Remote   
3752  105000             USD         105000                 US       Remote   

     company_location company_size    job_category  adjusted_salary  
3                  CA            M  Data Scientist       175000.000  
4                  CA            M  Data Scientist       120000.000  
7                  CA            M  Data Scientist       219000.000  
8                  CA            M  Data Scientist       141000.000  
9                  US            M  Data Scientist       147100.000  
...               ...          ...             ...              ...  
3741               US            M  Data Scientist       125670.000  
3742               US            M  Data Scientist       147342.750  
3746               IL            M  Data Scientist       129536.192  
3750               US            L  Data Scientist       438780.000  
3752               US            S  Data Scientist       111825.000  

[538 rows x 13 columns]

In [102]:
df_science.groupby('experience_level').agg({'adjusted_salary' : 'median'}).sort_values('adjusted_salary', ascending = False)

adjusted_salary
experience_level                         
Executive                      177167.500
Senior                         161880.000
Mid / Intermediete Level        93847.992
Entry Level                     74540.000

# SOAL No. 2
Show the number of records for each company_location, sorted descending

## JAWABAN No.2

In [103]:
df_2.groupby('company_location').size().sort_values(ascending = False)

company_location
US    1929
GB     155
CA      83
IN      57
DE      55
      ... 
MT       1
MD       1
PH       1
SK       1
VN       1
Length: 72, dtype: int64

# SOAL No.3
From this data, show the median adjusted salary of employee who works for company located in Indonesia, do you think that value represents the salary in Indonesia? why? why not?

## JAWABAN No.3

In [106]:
df_indo

work_year          experience_level employment_type       job_title  \
3061       2022  Mid / Intermediete Level       Full Time  Data Scientist   
3139       2022               Entry Level       Full Time    Data Analyst   

       salary salary_currency  salary_in_usd employee_residence remote_ratio  \
3061  4200000             INR          53416                 IN       Remote   
3139    15000             USD          15000                 ID      On-site   

     company_location company_size      job_category  adjusted_salary  
3061               ID            L    Data Scientist        58116.608  
3139               ID            L  Data Engineering        15975.000

In [105]:
#location based

df_indo = df_2[df_2['company_location'] == 'ID']
df_indo['adjusted_salary'].median()

np.float64(37045.804000000004)

In [107]:
#location based

df_global = df_2[df_2['company_location'] != 'ID']
df_global['adjusted_salary'].median()

np.float64(134107.0)

In [108]:
#experience - mid id

df_experience1 = df_indo[df_indo['experience_level'] == 'Mid / Intermediete Level']
df_experience1['adjusted_salary'].median()

np.float64(58116.60800000001)

In [109]:
#experience - mid global
df_experience2 = df_global[df_global['experience_level'] == 'Mid / Intermediete Level']
df_experience2['adjusted_salary'].median()

np.float64(99204.75)

In [110]:
#exp entry - id
df_entry = df_indo[df_indo['experience_level'] == 'Entry Level']
df_entry['adjusted_salary'].median()

np.float64(15975.0)

In [111]:
#exp entry - global
df_entry2 = df_global[df_global['experience_level'] == 'Entry Level']
df_entry2['adjusted_salary'].median()

np.float64(65000.0)

**Kesimpulan** <br>
kesimpulan berdasarkan  (tabel :#location based)
- indonesia : $37045 / thn -> Rp. 47.605.427 / bln
- global : $134107 / thn -> Rp. 167.633.750 / bln

Berdasarkan data BPS pada tahun 2023, median penghasilan kelas pekerja di indonesia adalah Rp. 2.500.000. (via x @hrdbacot)

Dilansir situs jobstreet Indonesia pada bulan agustus tahun 2024 range gaji seorang data scientist di Indonesia sebesar Rp. 8.600.000 - Rp. 11.600.000

Berdasarkan situs lain glassdoor.com seorang senior data scientist di Indonesia berkisar antara Rp.17.000.000 - Rp.32.000.000 perbulan

Dari hasil 3 penelusuran data tersebut, dapat disimpulkan kalau data diatas tidak merpresentasikan nilai gaji pekerja di Indonesia.

kesimpulan tambahan berdasarkan lokasi dan experience level (tabel #experience)
- mid : indo : $58116 / thn -> Rp. 74.785.606 / bln
- entry : indo : $15975 / thn -> Rp. 20.557.162 / bln

Hasil yang sama juga didapatkan apabila membreakdown gaji dengan lokasi di Indonesia berdasarkan level experience. Dimana data diatas masih jauh lebih tinggi jika dibandingkan dengan realita gaji di Indonesia